### ISSUE TITLE
[remote] Support Nextstrain Groups as remote destinations using nextstrain.org credentials

### ISSUE THREAD

**Issue by tsibley** [nextstrain/cli#90]
Currently the `nextstrain remote` family of commands only support `s3://` remote destinations, but they're designed with the intention to support other kinds of remote destinations in the future. The future is now. :-)

One of the pain points with Nextstrain Groups is that using `nextstrain remote` to deploy, list, and manage datasets in your group requires setting up AWS credentials and knowing the name/path of the S3 bucket. Instead, we'd like to support using nextstrain.org credentials (see #89) and a custom Nextstrain Groups destination syntax, thus decoupling the deploy process from our AWS implementation.

What should the destination syntax be? I can imagine the following, in my rough order of preference:

```
nextstrain remote upload nextstrain.org/groups/blab/ auspice/sars-like-cov*.json
nextstrain remote upload nextstrain:groups/blab/ auspice/sars-like-cov*.json
nextstrain remote upload groups:blab auspice/sars-like-cov*.json
nextstrain remote upload groups/blab auspice/sars-like-cov*.json
```

This draws on Docker's image syntax for inspiration. The nice parts of the first syntax are that a) it matches how groups are accessed on the web and b) supports local testing of the nextstrain.org server component by making the remote destination, e.g. `localhost:4000/groups/blab`. The above also aren't all mutually exclusive if we want to support shorter aliases for the longer, more general form.

#89 is a strict prerequisite. Changes to nextstrain.org's web API (what I sometimes call Charon v2) would also be really nice so that the CLI can talk to it and be unaware of AWS implementation details. (I also see such a Charon v2 API as providing a foundation for better static hosting of Auspice.) However, interim solutions exist where the CLI maps remotes for Nextstrain Groups to S3 buckets and talks to S3 behinds the scenes using temporary credentials granted by Cognito.

**Comment by trvrb** [comment]
This is a really great direction. I also prefer (1) from an imagined user perspective. This cuts out redirection and makes it clear where the files are going.

**Comment by jameshadfield** [comment]
Looks great! I also agree that syntax 1 seems the best.

> we'd like to support using nextstrain.org credentials

Currently we have read-only nextstrain.org credentials (behind the scenes these are implemented as cognito users) and write credentials (an IAM user who can modify the contents of the S3 bucket). This gives a clear separation between those who can modify data and those who can just view data, at the expense of maintaining separate credentials. Is the idea here that _some_ nextstrain.org credentials (i.e. cognito users) will get write access? (e.g. via an additional cognito group or similar). Happy for this conversation to be moved to #89 if that's more appropriate.

re: adding nextstrain.org API endpoints

This would be great as it paves the way for a web-UI layer sometime in the future. Could you list a few examples of what you're thinking of here? (I remember we discussed this on slack at some point)

**Comment by tsibley** [comment]
> This gives a clear separation between those who can modify data and those who can just view data, at the expense of maintaining separate credentials. Is the idea here that some nextstrain.org credentials (i.e. cognito users) will get write access? (e.g. via an additional cognito group or similar).

This is an unintended separation that falls out of the current implementation's shortcomings. We hadn't aimed for different access levels within a group initially, but it is certainly something we can implement with time. Levels like read/write/admin might be reasonable, for example, and can be implemented a number of ways (additional Cognito groups behind the scenes being one way). 

The main reason I want to make the "Charon v2" API is so that nextstrain.org mediates access to Groups data, which then lets us easily make authorization decisions like access level at that layer. 

> re: adding nextstrain.org API endpoints
> 
> This would be great as it paves the way for a web-UI layer sometime in the future. Could you list a few examples of what you're thinking of here?

Exactly. The CLI and web UI would use the same web API. My thoughts on the v2 API boil down to a typical set of RESTful endpoints that make use of what HTTP's good at. Some more detail is sketched out in these notes:

- https://github.com/tsibley/blab-standup/blob/master/2018-12-26.md#proposed
- https://github.com/tsibley/blab-standup/blob/master/2020-04-09.md

(check the date on the first link :scream:)

We could consider an alternate design in which the CLI and web UI still talk directly to AWS, but do so via transient credentials from Cognito-granted roles. There are some benefits there, but a downside is it cuts nextstrain.org out of the equation and limits some implementation options.

**Comment by tsibley** [comment]
If we're assuming the CLI will talk to nextstrain.org's Charon v2 API (which is not yet implemented or fully-designed), then this issue is just the CLI's half of the functionality. The other half will be in nextstrain.org's repo. Most of the CLI's half can be developed first by either commenting out the ultimate HTTP requests to nextstrain.org or stubbing them to something like http://httpbin.org (which run locally too).

**Comment by tsibley** [comment]
Resolved by #115  and https://github.com/nextstrain/cli/pull/141.
